synchronization/futex.rs

```diff
@@ -0,0 +1,524 @@
+        }
+    }
+
+    fn try_lock(&self) -> bool {
+        // Attempt to do the baseline atomic transaction.
+        let value = (self.0).compare_and_swap(0, gettid(), Ordering::SeqCst);
+
+        // If this is not equal to 0, we failed to lock.
+        return value == 0;
+    }
+}
+
+impl<T: RawFutex> Drop for FutexMutex<T> {
+    fn drop(&mut self) {
+        if likely(!std::thread::panicking()) {
+            // if we're already panicking I guess we don't need to panic again
+            // TODO make this log an error instead
+
+            if unlikely(self.islocked()) {
+                panic!("Trying to drop a locked mutex {:p}", &self)
+            }
+        }
+    }
+}
+
+pub struct FutexCondition(Futex);
+
+impl FutexCondition {
+    fn inner_lock(&self, mutex: &FutexMutex<Futex>) {
+        let mutex_result = mutex.mutex_get(false);
+        if mutex_result == MutexLockResult::Ok {
+            // successfully locked
+            return;
+        } else {
+            panic!("Did not expect an error while locking a condition variable mutex");
+        }
+    }
+
+    pub fn wait(&self, mutex: &FutexMutex<Futex>) {
+        let start = self.0.load(Ordering::SeqCst);
+
+        mutex.unlock();
+
+        loop {
+            let ret = self.0.wait(start, None);
+
+            match ret {
+                Ok(_) => {
+                    // we need to grab the lock.
+                    return self.inner_lock(mutex);
+                }
+                Err(NixError::Sys(errno)) => {
+                    if unlikely(errno == Errno::EAGAIN) {
+                        return self.inner_lock(mutex);
+                    }
+
+                    if unlikely(errno == Errno::EINTR) {
+                        continue;
+                    }
+
+                    panic!("Unexpected failure while waiting on futex: FUTEX_WAIT({:p}, {}, None) failed due to {}", &self, start, errno);
+                }
+                _ => {
+                    unreachable!("Should never get this error");
+                }
+            }
+        }
+    }
+
+    pub fn wake(&self, _mutex: &FutexMutex<Futex>, number: WakeNumber) -> i32 {
+        self.0.fetch_add(1, Ordering::SeqCst);
+
+        self.0.wake(number).expect("Failed to wake CV")
+    }
+}
+
+pub type Mutex<T> = lock_api::Mutex<FutexMutex<Futex>, T>;
+pub type MutexGuard<'a, T> = lock_api::MutexGuard<'a, FutexMutex<Futex>, T>;
+
+/// Wrapper around lower level synchronization primitives allowing
+/// `std::sync::CondVar`-like usage which can be sent through shared
+/// memory and work across multiple properties.
+pub struct CondVar<T> {
+    condition: FutexCondition,
+    mutex: Mutex<T>,
+}
+
+impl<T> CondVar<T> {
+    /// Returns a new condition variable that works with the mutex passed in.
+    pub fn new(mutex: Mutex<T>) -> CondVar<T> {
+        CondVar {
+            condition: FutexCondition(Futex(AtomicI32::new(0))),
+            mutex,
+        }
```

Comment:
Is it safe to just take in the mutex we're given, given that it could already be acquired (and thus acquired when the condvar initializes, where I assume it assumes that the mutex is unlocked to begin with)?

Reply:
No that would be a multiple borrow. When a mutex is acquired it gets borrowed into a MutexGuard, which unlocks the mutex when it drops. Note that this is our `Mutex` type not our `RawMutex` type.